File: plugins/ghostwriter/agent/tools/detector/lib/threshold.py
"""
threshold.py

Detection-specific threshold checking logic.
Retrieves rule configs and compares measured values against AI thresholds.
"""

from typing import Any, Dict, Optional


def get_threshold(
    category: str,
    rule: str,
    config: Dict[str, Any]
) -> Optional[Dict[str, Any]]:
    """
    Get threshold configuration for a rule.

    Args:
        category: Category name (e.g., "punctuation", "burstiness")
        rule: Rule name (e.g., "period_quote", "coefficient_of_variation")
        config: Resolved writing configuration

    Returns:
        ThresholdConfig dict or None if rule is disabled
    """
    rules = config.get("rules", {})
    category_config = rules.get(category, {})
    rule_config = category_config.get(rule)

    if rule_config and rule_config.get("disabled"):
        return None

    return rule_config


def check_threshold(
    category: str,
    rule: str,
    value: float,
    config: Dict[str, Any],
    direction: str = "above"
) -> Optional[Dict[str, Any]]:
    """
    Check if a rule should trigger an AI signal.

    Args:
        category: Category name (e.g., "punctuation", "burstiness")
        rule: Rule name (e.g., "period_quote", "coefficient_of_variation")
        value: The measured value to check against the threshold
        config: Resolved writing configuration
        direction: "above" if value should be above threshold to signal AI,
                   "below" if value should be below threshold to signal AI

    Returns:
        Dict with shouldFlag and reason, or None if rule disabled
    """
    threshold = get_threshold(category, rule, config)

    if threshold is None:
        return None

    ai_threshold = threshold.get("ai_threshold")

    if ai_threshold is None:
        media = config.get("publication", {}).get("media", "unknown")
        return {"shouldFlag": False, "reason": f"{category}.{rule} unreliable for {media}"}

    if direction == "above":
        should_flag = value > ai_threshold
    else:
        should_flag = value < ai_threshold

    media = config.get("publication", {}).get("media", "unknown")
    reason = (
        f"{category}.{rule}: {value:.2f} {'>' if direction == 'above' else '<'} {ai_threshold} (AI threshold for {media})"
        if should_flag
        else ""
    )

    return {"shouldFlag": should_flag, "reason": reason}

File: src/plugins/ghostwriter/agent/tools/detector/lib/threshold.ts
/**
 * threshold.ts
 *
 * Detection-specific threshold checking logic.
 * Retrieves rule configs and compares measured values against AI thresholds.
 */

import type {
  ResolvedWritingConfig,
  BaseRuleConfig,
  RuleCategory,
} from '../../config/schemas';

/**
 * Get threshold configuration for a rule from a resolved config.
 *
 * Retrieves the configuration object for a specific rule within a rule category.
 * Returns null if the rule is disabled or doesn't exist for the given config context.
 *
 * @param category - Rule category (e.g., "vocabulary", "punctuation", "structure")
 * @param rule - Rule name within the category (e.g., "ai_vocab_density")
 * @param config - Resolved writing config
 * @returns BaseRuleConfig object with thresholds, or null if rule is disabled or not found
 */
export function getThreshold(
  category: RuleCategory,
  rule: string,
  config: ResolvedWritingConfig,
): BaseRuleConfig | null {
  // Get the category configuration
  const categoryConfig = config.rules[category];
  if (!categoryConfig) {
    return null;
  }

  // Get the rule configuration (use type assertion due to complex nested types)
  const ruleConfig = (
    categoryConfig as Record<string, BaseRuleConfig | undefined>
  )[rule];
  if (!ruleConfig || ruleConfig.disabled) {
    return null;
  }

  return ruleConfig;
}

/**
 * Check if a rule should trigger an AI signal based on config.
 *
 * Compares a measured value against the AI threshold for a rule and determines
 * whether it signals AI-generated content. Returns null if the rule is disabled
 * for the given context.
 *
 * The direction parameter controls the comparison:
 * - "above": signals AI if value > ai_threshold (e.g., RepetitionRate > 0.8)
 * - "below": signals AI if value < ai_threshold (e.g., Perplexity < 20)
 *
 * @param category - Rule category (e.g., "vocabulary", "punctuation", "structure")
 * @param rule - Rule name within the category (e.g., "ai_vocab_density")
 * @param value - Measured value to check against threshold
 * @param config - Resolved writing config
 * @param direction - Comparison direction: "above" (default) or "below"
 * @returns Object with `shouldFlag` (boolean) and `reason` (string), or null if rule
 *          is disabled or unreliable for the given context
 */
export function checkThreshold(
  category: RuleCategory,
  rule: string,
  value: number,
  config: ResolvedWritingConfig,
  direction: 'above' | 'below' = 'above',
): { shouldFlag: boolean; reason: string } | null {
  // Get the threshold
  const threshold = getThreshold(category, rule, config);

  // Rule disabled for this context
  if (!threshold) {
    return null;
  }

  // Threshold is null = unreliable for this context
  if (threshold.ai_threshold === null || threshold.ai_threshold === undefined) {
    return {
      shouldFlag: false,
      reason: `${category}.${rule} unreliable for ${config.publication.media}`,
    };
  }

  // Auto-disable when human baseline exceeds AI threshold (rule not useful)
  if (threshold.human_baseline !== undefined && threshold.human_baseline !== null) {
    if (direction === 'above' && threshold.human_baseline >= threshold.ai_threshold) {
      return null;
    }
    if (direction === 'below' && threshold.human_baseline <= threshold.ai_threshold) {
      return null;
    }
  }

  const shouldFlag =
    direction === 'above'
      ? value > threshold.ai_threshold
      : value < threshold.ai_threshold;

  const reason = shouldFlag
    ? `${category}.${rule}: ${value.toFixed(2)} ${direction === 'above' ? '>' : '<'} ${threshold.ai_threshold} (AI threshold for ${config.publication.media})`
    : '';

  return { shouldFlag, reason };
}
